### DISCUSSION TITLE
Display current node from selection on the dropdown menu

### DISCUSSION THREAD

**Discussion by augustomallmann** [ueberdosis/tiptap#3856] [General Questions]
Hey, hope you're well.

I'm developing an application with Tiptap v2 and I'm trying to reproduce [this dropdown menu](https://playground.lexical.dev/) from Lexical: 

<img width="465" alt="image" src="https://user-images.githubusercontent.com/8194056/225399784-75c4797a-efb7-42bd-af5e-143d2acb9157.png">

It display the current node from the selection. Example, if you click on a `quote`, it displays the `quote` node as active on the dropdown button. [Github code.
](https://github.com/facebook/lexical/blob/main/packages/lexical-playground/src/plugins/ToolbarPlugin/index.tsx#L635)

I'm trying to achieve this same behavior by getting the current selected node as the code below:

```
editor.on('selectionUpdate', ({ editor }) => {
    const currentElementSelectedType = editor.state.selection.$head.parent.type.name;
})
```

What I've noticed is that it is making the editor to freeze. My dropdown button component is the following: 

```const DropdownButton = () => {

  const {  
    activeItem, 
    items 
  } = useEditorDropdown();

  return (
    <Flex 
      align="center" 
      className="toolbar_dropdown_select_button"
    >
      <FlexItem 
        className="toolbar_dropdown_select_button_icon" 
        displayFlex
        key={activeItem ? activeItem.icon : items[0].icon}
      >
        <Icon icon={activeItem ? activeItem.icon : items[0].icon} size="sm" />
      </FlexItem>
      <FlexItem className="toolbar_dropdown_select_button_text">
        {activeItem ? activeItem.text : items[0].text}
      </FlexItem>
    </Flex>
  );
};
```

It displays the node that was last selected by the user from the `DropdownList`:

```
onst DropdownItems = () => {

  const { dropdownPlacement, isOpen, items, setIsOpen, setActiveItem } = useEditorDropdown();

  const handleClickItem = (
    index: number, 
    command: () => boolean
  ) => {
    command()
    setIsOpen(false)
    setActiveItem({
      icon: items[index].icon,
      text: items[index].text,
    })
  }

  return (
    isOpen && (
      <Card
        className={`toolbar_dropdown_list_card ${dropdownPlacement ? `placement_${dropdownPlacement}` : ""}`}
        paddingX="none"
        paddingY="xs"
        shadow="deeper"
      >
        <Nav variant="subtle">
          {items.map(({ icon, text, command, isActive }, index) => (
            <NavItem
              className={`toolbar_dropdown_list_item ${isActive ? "active" : ""}`}
              iconLeft={icon}
              key={`${text}_${index}`}
              margin="none"
              onClick={() => handleClickItem(index, command)}
              text={text}
            />
          ))}
        </Nav>
      </Card>
    )
  );
};
```

The list of items is rendered from the following array:

```
const toolbarDropdownItems = (editor: Editor) => [
    {
        node: "paragraph",
        icon: "paragraph",
        isActive: editor.isActive("paragraph"),
        text: "Paragraph",
        command: () => paragraphCommand(editor),
    },
    {
        node: "heading",
        icon: "h1",
        isActive: editor.isActive("heading", {level: 1}),
        text: "Heading 1",
        command: () => headingCommand(editor, 1),
    },
    {
        node: "heading",
        icon: "h2",
        text: "Heading 2",
        command: () => headingCommand(editor, 2),
    },
    {
        node: "heading",
        icon: "h3",
        text: "Heading 3",
        command: () => headingCommand(editor, 3),
    },
    {
        node: "heading",
        icon: "h4",
        text: "Heading 4",
        command: () => headingCommand(editor, 4),
    },
    {
        // node: "heading",
        icon: "list",
        isActive: editor.isActive("bulletList"),
        text: "Bullet List",
        command: () => bulletListCommand(editor),
    },
    {
        icon: "list-ol",
        text: "Ordered List",
        command: () => orderedListCommand(editor),
    },
    {
        icon: "block-quote",
        text: "Block Quote",
        command: () => blockquoteCommand(editor),
    },
]
```

Questions:

- How do I know which is the current node selected from the user?
- When using `editor.state.selection.$head.parent.type.name` to get the node name, if its a list, it says that the selected element is a paragraph. How to identify a list?

**Comment by johnpuddephatt** [comment]
> When using editor.state.selection.$head.parent.type.name to get the node name, if its a list, it says that the selected element is a paragraph. How to identify a list?

You're most likely seeing that the selected element is a paragraph because that is the current node name – in tiptap/prosemirror list items have paragraphs in them, e.g.

```
<ul>
    <li>
        <p>List item</p>
    </li>
</ul>
```

I am definitely not a Prosemirror (or Tiptap!) expert but I think to get the 'top level' element you can do:
```
this.editor.view.state.selection.$head.node(1)
```
The number provided indicates the level you want, with 0 being the editor root node, 1 being the first level child, etc. I was trying to do something similar in [an editor I was playing around with here](http://gutentap.letsdance.agency/) where I wanted to display the currently selected 'block'.
